File: terapix/youpi/plugins/skeleton.py
##############################################################################
#
# Copyright (c) 2008-2009 Terapix Youpi development team. All Rights Reserved.
#                    Mathias Monnerville <monnerville@iap.fr>
#                    Gregory Semah <semah@iap.fr>
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
##############################################################################

# vim: set ts=4

#
# Mandatory data members
#
# def __getCondorSubmissionFile(self, request)		: generates a suitable Condor submission file
# def getResultEntryDescription(self, task)			: returns custom result entry description for a task.
# def getSavedItems(self, request)					: returns a user's saved items for this plugin 
# def getTaskInfo(self, request)					: returns information about a finished processing task. Used on the results page.
# def process(self, request)						: generates and submits a Condor submission file (CSF)
# def reprocessAllFailedProcessings(self, request)	: returns parameters to allow reprocessing of failed processings
# def saveCartItem(self, request)					: save cart item's custom data to DB
#

import sys, os.path, re, time, string
import marshal, base64, zlib
#
from terapix.youpi.pluginmanager import ProcessingPlugin
from terapix.exceptions import *
from terapix.youpi.models import *
from terapix.youpi.auth import read_proxy
import terapix.lib.cluster.condor as condor
#
from django.conf import settings

class Skeleton(ProcessingPlugin):
	"""
	This plugin does nothing but can serve as a base for writing processing plugins
	"""
	def __init__(self):
		ProcessingPlugin.__init__(self)

		self.id = 'skel'
		self.optionLabel = 'Skeleton (DEMO)'
		self.description = 'Skeleton DEMO processing'
		# Item prefix in processing cart. This should be short string since
		# the item ID can be prefixed by a user-defined string
		self.itemPrefix = 'SKEL'
		self.index = 65535

		self.template = 'plugins/skeleton.html' 						# Main template, rendered in the processing page
		self.itemCartTemplate = 'plugins/skeleton_item_cart.html' 		# Template for custom rendering into the processing cart
		self.jsSource = 'plugins/skeleton.js' 							# Custom javascript

		# Will queue jobCount jobs on the cluster
		self.jobCount = 5
		self.command = '/usr/bin/uptime'

	def process(self, request):
		"""
		Do the job.
		1. Generates a condor submission file
		2. Executes condor_submit on that file
		3. Returns info related to ClusterId and number of jobs submitted
		"""
		post = request.POST

		cluster_ids = []
		error = condorError = '' 

		try:
			csfPath = self.__getCondorSubmissionFile(request)
			pipe = os.popen(os.path.join(settings.CONDOR_BIN_PATH, "condor_submit %s 2>&1" % csfPath))
			data = pipe.readlines()
			pipe.close()
			cluster_ids.append(self.getClusterId(data))
		except CondorSubmitError, e:
			condorError = str(e)
		except Exception, e:
			error = "Error while processing item: %s" % e

		return {'ClusterIds': cluster_ids, 'Error': error, 'CondorError': condorError}

	def __getCondorSubmissionFile(self, request):
		"""
		Generates a suitable Condor submission file for processing self.command jobs on the cluster.
		"""

		post = request.POST
		try:
			#
			# Retreive your POST data here
			#
			resultsOutputDir = post['ResultsOutputDir']
			itemId = str(post['ItemId'])
		except Exception, e:
				raise PluginError, "POST argument error. Unable to process data: %s" % e

		now = time.time()
		# Condor submission file
		csfPath = condor.CondorCSF.getSubmitFilePath(self.id)
		csf = open(csfPath, 'w')

		# Content of YOUPI_USER_DATA env variable passed to Condor
		# At least those 3 keys
		userData = {'Descr' 			: str("%s trying %s" % (self.optionLabel, self.command)),		# Mandatory for Active Monitoring Interface (AMI)
					'Kind'	 			: self.id,														# Mandatory for AMI, Wrapper Processing (WP)
					'UserID' 			: str(request.user.id),											# Mandatory for AMI, WP
					'ItemID' 			: itemId, 
					'ResultsOutputDir'	: str(resultsOutputDir)											# Mandatory for WP
				} 

		#
		# Generate CSF
		#
		cluster = condor.YoupiCondorCSF(request, self.id, desc = self.optionLabel)

		# Real command to perform here
		args = ''
		for x in range(self.jobCount):
			# Mandatory for WP
			userData['JobID'] = self.getUniqueCondorJobId()
			encUserData = base64.encodestring(marshal.dumps(userData)).replace('\n', '')
			cluster.addQueue(
				queue_args = str("%(data)s %(command)s" % {'data': encUserData, 'command': self.command,}),
				queue_env = {'YOUPI_USER_DATA': encUserData,}
			)

		cluster.write(csfPath)
		return csfPath

	def getTaskInfo(self, request):
		"""
		Returns information about a finished processing task. Used on the results page.
		"""

		post = request.POST
		try:
			taskid = post['TaskId']
		except Exception, e:
			raise PluginError, "POST argument error. Unable to process data."

		task, filtered = read_proxy(request, Processing_task.objects.filter(id = taskid))
		if not task:
			return {'Error': str("Sorry, you don't have permission to see this result entry.")}
		task = task[0]

		# Error log content
		if task.error_log:
			err_log = str(zlib.decompress(base64.decodestring(task.error_log)))
		else:
			err_log = ''

		return {	'TaskId'	: str(taskid),
					'Title' 	: str("%s" % self.description),
					'User' 		: str(task.user.username),
					'Success' 	: task.success,
					'ClusterId'	: str(task.clusterId),
					'Start' 	: str(task.start_date),
					'End' 		: str(task.end_date),
					'Duration' 	: str(task.end_date-task.start_date),
					'Log' 		: err_log
			}

	def getResultEntryDescription(self, task):
		"""
		Returns custom result entry description for a task.
		task: django object

		returned value: HTML tags allowed
		"""

		return "%s <tt>%s</tt>" % (self.optionLabel, self.command)

	def saveCartItem(self, request):
		"""
		Save cart item's custom data to DB
		"""

		post = request.POST
		try:
			itemID = str(post['ItemId'])
			resultsOutputDir = post['ResultsOutputDir']
		except Exception, e:
			raise PluginError, "POST argument error. Unable to process data."

		items = CartItem.objects.filter(kind__name__exact = self.id).order_by('-date')
		if items:
			itemName = "%s-%d" % (itemID, int(re.search(r'.*-(\d+)$', items[0].name).group(1))+1)
		else:
			itemName = "%s-%d" % (itemID, len(items)+1)

		# Custom data
		data = { 'Descr' : "Runs %d %s commands on the cluster" % (self.jobCount, self.command),
				 'resultsOutputDir' : resultsOutputDir }
		sdata = base64.encodestring(marshal.dumps(data)).replace('\n', '')

		profile = request.user.get_profile()
		k = Processing_kind.objects.filter(name__exact = self.id)[0]
		cItem = CartItem(kind = k, name = itemName, user = request.user, mode = profile.dflt_mode, group = profile.dflt_group)
		cItem.data = sdata
		cItem.save()

		return "Item %s saved" % itemName

	def reprocessAllFailedProcessings(self, request):
		"""
		Returns parameters to allow reprocessing of failed processings
		"""
		# FIXME: TODO
		pass

	def getSavedItems(self, request):
		"""
		Returns a user's saved items for this plugin 
		"""
		items, filtered = read_proxy(request, CartItem.objects.filter(kind__name__exact = self.id).order_by('-date'))
		res = []
		for it in items:
			data = marshal.loads(base64.decodestring(str(it.data)))
			res.append({'date' 				: "%s %s" % (it.date.date(), it.date.time()), 
						'username'			: str(it.user.username),
						'descr' 			: str(data['Descr']),
						'itemId' 			: str(it.id), 
						'resultsOutputDir' 	: str(self.getUserResultsOutputDir(request)),
						'name' 				: str(it.name) })
		return res

File: terapix/templates/plugins/skeleton.js
/*****************************************************************************
 *
 * Copyright (c) 2008-2009 Terapix Youpi development team. All Rights Reserved.
 *                    Mathias Monnerville <monnerville@iap.fr>
 *                    Gregory Semah <semah@iap.fr>
 *
 * This program is Free Software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 *****************************************************************************/

/* vim: set ts=4 */

/*
 * Group: Skeleton Plugin
 *
 * JS code for Skeleton plugin.
 *
 * Functions in this file are required for every processing plugin to work. Feel free
 * to add any content in their bodies, but keep the declarations!
 *
 * Mandatory function prototypes:
 *
 * addSelectionToCart()						: adds current selection to cart
 * addToCart(data)							: adds a previously saved selection to cart
 * delSavedItem(trid, name)					: deletes a previously saved item
 * renderOutputDirStats(container_id)		: renders some statistics (from the results page)
 * reprocessAllFailedProcessings(tasksList)	: one click for reprocessing failed processings from the results page
 * resultsShowEntryDetails(container_id)	: displays custom results information on the results page
 * run(trid, opts, silent)					: submits a job to the cluster
 * saveItemForLater(trid, opts, silent)		: save item for later processing then remove item from the shopping cart
 * showSavedItems()							: displays saved cart items
 *
 */

/* global */
var uidskel = '{{ plugin.id }}';

var {{ plugin.id }} = {
	/*
	 * Function: addSelectionToCart
	 * Add the current selection to cart.
	 * Since this is a demo plugin, no real selection has been made. It's just adding the output data 
	 * directory to the userData variable
	 *
	 */
	addSelectionToCart: function() {
		// Do your sanity checks here
		var msg = 'One Skeleton-related job has been added to the cart. This is a DEMO plugin: it does nothing useful.';
	
		// Custom output directory
		var output_data_path = '{{ processing_output }}{{ user.username }}/' + uidskel + '/';
	
		// Set mandatory structures
		var p_data = {	
			plugin_name : uidskel,
			userData : {resultsOutputDir: output_data_path}
		};
	
		// Add entry into the shopping cart
		s_cart.addProcessing(	
			p_data,
			// Custom handler
			function() {
				document.fire('notifier:notify', msg);
			}
		);
	},

	/*
	 * Function: run
	 * Submits a job to the cluster
	 *
	 * Parameters:
	 *  trid - string: for row number
	 *  opts - hash: options
	 *  silent - boolean: silently submit item to the cluster
	 *
	 */ 
	run: function(trid, opts, silent) {
		var silent = typeof silent == 'boolean' ? silent : false;
		var runopts = get_runtime_options(trid);
		var logdiv = $('master_condor_log_div');
	
		var r = new HttpRequest(
				logdiv,
				null,	
				// Custom handler for results
				function(resp) {
					r = resp.result;
					var success = update_condor_submission_log(resp, silent);
					if (!success) return;
	
					// Silently remove item from the cart
					removeItemFromCart(trid, true);
				}
		);
	
		opts = $H(opts);
		opts.set('Plugin', uidskel);
		opts.set('Method', 'process');
		opts.set('ReprocessValid', (runopts.reprocessValid ? 1 : 0));
		opts = opts.merge(runopts.clusterPolicy.toQueryParams());

		r.send('/youpi/process/plugin/', opts.toQueryString());
	},

	/*
	 * Function:  reprocessAllFailedProcessings
	 * One click for reprocessing failed processings from the results page
	 *
	 * Parameters:
	 *  taskList - array: list of task Ids
	 *
	 */ 
	reprocessAllFailedProcessings: function(tasksList) {
		alert('TODO...');
	},

	/*
	 * Function: renderOutputDirStats
	 * Renders some statistics (from the results page)
	 * Those stats are bound to the results output directory.
	 *
	 * Parameters:
	 *  container_id - string: DOM container id
	 *
	 */ 
	renderOutputDirStats: function(container_id) {
		var container = $(container_id).update();
	
		// global var defined in results.html
		var resp = output_dir_stats_data;
		var stats = resp.Stats;
	
		var tab = new Element('table').addClassName('output_dir_stats');
		var tr,th,td;
		var tr = new Element('tr');
		// Headers
		var headers = $H({	
			task_success: 'Task success',
			task_failure: 'Task failures',
			task_total: 'Total processings'
		});
		headers.each(function(header) {
			th = new Element('th', {'colspan': '2'}).addClassName(header.key).update(header.value);
			tr.insert(th);
		});
		tab.insert(tr);
	
		tr = new Element('tr');
		var val, percent, cls;
		headers.each(function(header, k) {
			switch (k) {
				case 0:
					val = stats['TaskSuccessCount'][0];
					percent = stats['TaskSuccessCount'][1] + '%';
					cls = 'task_success';
					break;
				case 1:
					val = stats['TaskFailureCount'][0];
					percent = stats['TaskFailureCount'][1] + '%';
					cls = 'task_failure';
					break;
				case 2:
					val = stats['Total'];
					percent = '100%';
					cls = 'task_total';
					break;
				default:
					break;
			}
			var td = new Element('td').addClassName(cls).update(val);
			tr.insert(td);

			td = new Element('td').addClassName(cls).update(percent);
			tr.insert(td);
		});
		tab.insert(tr);
		container.insert(tab);
	},

	/*
	 * Function: saveItemForLater
	 * Save item for later processing then remove item from the shopping cart
	 *
	 * Parameters:
	 *  trid - string: for row number
	 *  opts - hash: options
	 *  silent - boolean: silently submit item to the cluster
	 *
	 */ 
	saveItemForLater: function(trid, opts, silent) {
		opts = $H(opts);
		opts.set('Plugin', uidskel);
		opts.set('Method', 'saveCartItem');

		var runopts = get_runtime_options(trid);
		var r = new HttpRequest(
				uidskel + '_result',
				null,	
				// Custom handler for results
				function(resp) {
					document.fire('notifier:notify', 'Skeleton item saved successfully');
					// Silently remove item from the cart
					removeItemFromCart(trid, true);
				}
		);
	
		r.send('/youpi/process/plugin/', opts.toQueryString());
	},

	/*
	 * Function: resultsShowEntryDetails
	 * Displays custom results information on the results page. 
	 * 'resp' contains server-side info to display
	 *
	 * Parameters:
	 *  container_id - string: id of DOM container
	 *
	 */ 
	resultsShowEntryDetails: function(container_id) {
		var tr, th, td;
		// See templates/results.html, function showDetails(...)
		var resp = currentReturnedData;
		var container = $(container_id);
		var d = new Element('div').addClassName('entryResult');
		var tab = new Element('table').addClassName('fileBrowser').setStyle({width: '100%'});
	
		tr = new Element('tr');
		th = new Element('th').update(resp['Title']);
		tr.insert(th);
		tab.insert(tr);
	
		// Duration
		var tdiv = new Element('div').addClassName('duration');
		tdiv.insert(resp.Start + '<br/>');
		tdiv.insert(resp.End + '<br/>');
		var src = resp.Success ? 'success' : 'error';
		var img = new Element('img', {src: '/media/themes/{{ user.get_profile.guistyle }}/img/admin/icon_' + src + '.gif'}).setStyle({paddingRight: '5px'});
		tdiv.insert(img);
		tdiv.insert(resp.Duration);
		tr = new Element('tr');
		td = new Element('td').setStyle({borderBottom: '2px #5b80b2 solid'});
		td.insert(tdiv);
		tr.insert(td);
		tab.insert(tr);
	
		// User
		var udiv = new Element('div').addClassName('user');
		udiv.insert('Job initiated by ' + resp.User);
		udiv.insert(new Element('br'));
		udiv.insert('Exit status: ');
		udiv.insert(new Element('br'));
		var txt = resp.Success ? 'success' : 'failure';
		var exit_s = new Element('span').addClassName('exit_' + txt).update(txt);
		udiv.insert(exit_s);
		td.insert(udiv);
	
		tr = new Element('tr');
		td = new Element('td', {style: 'padding: 0px'});
		var tab2 = new Element('table').addClassName('qfits-result-entry-params');
		td.insert(tab2);
		tr.insert(td);
		tab.insert(tr);
	
		// Error log file when failure
		if (!resp.Success) {
			tr = new Element('tr');
			td = new Element('td', {colspan: 2}).setStyle({borderBottom: '2px #5b80b2 solid'});
			var but = new Element('input', {type: 'button',
											value: 'Toggle error log file view',
											onclick: "toggleDisplay('log-" + resp.TaskId + "');"
			});
			td.insert(but);
			var cdiv = new Element('div', {id: 'log-' + resp.TaskId}).setStyle({
				height: '200px', 
				overflow: 'auto', 
				backgroundColor: 'black', 
				paddingLeft: '5px'
			}).hide();
			var pre = new Element('pre').update(resp.Log);
			cdiv.insert(pre);
			td.insert(cdiv);
			tr.insert(td);
			tab2.insert(tr);
		}
		
		// Permissions
		tr = new Element('tr');
		td = new Element('td').setStyle({padding: '0px'});
		td.update(ResultsHelpers.getPermissionsEntry(resp.TaskId));
		tr.insert(td);
		tab2.insert(tr);

		// Condor Job Logs
		tr = new Element('tr');
		td = new Element('td').setStyle({padding: '0px'});
		td.update(ResultsHelpers.getCondorJobLogsEntry(resp.ClusterId, resp.TaskId));
		tr.insert(td);
		tab2.insert(tr);
	
		d.insert(tab);
		container.insert(d);
	},

	/*
	 * Function: showSavedItems
	 * Displays saved cart items
	 *
	 */ 
	showSavedItems: function() {
		var cdiv = $('plugin_menuitem_sub_' + uidskel).update();
		var div = new Element('div', {id: uidskel + '_saved_items_div'}).addClassName('savedItems');
		cdiv.insert(div);
	
		var r = new HttpRequest(
				div.id,
				null,	
				// Custom handler for results
				function(resp) {
					div.update();
	
					var total = resp.result.length;
					if (total == 0) {
        	            div.update('No saved item');
						return;
					}
					var table = new Element('table').addClassName('savedItems');
					var tr, th;

					tr = new Element('tr');
					var header = $A(['Date', 'Permissions', 'Name', 'Description', 'Action']);
					header.each(function(elem) {
						tr.insert(new Element('th').update(elem));
					});
					table.insert(tr);
	
					var delImg, trid;
					var tabi, tabitr, tabitd;
					resp.result.each(function(res, k) {
						trid = uidskel + '_saved_item_' + k + '_tr';
						tr = new Element('tr', {id: trid});
						// Delete
						delImg = new Element('img', {	id: uidskel + '_del_saved_item_' + k,	
														src: '/media/themes/{{ user.get_profile.guistyle }}/img/misc/delete.gif'
						}).setStyle({marginRight: '5px'}).hide();
						delImg.c_data = {trid: trid, name: res.name};
						delImg.observe('click', function() {
							{{ plugin.id }}.delSavedItem(this.c_data.trid, this.c_data.name);
						});
	
						// Date
						td = new Element('td').update(res.date);
						tr.insert(td);
	
						// Permissions
						td = new Element('td').addClassName('config').update(get_permissions('cartitem', res.itemId, function(r, imgId) {
							// imgId is the misc parameter passed to get_permissions()
							var img = $(imgId);
							r.currentUser.write ? img.show() : img.hide();
						}, delImg.readAttribute('id') /* Misc data */));
						tr.insert(td);
	
						// Name
						td = new Element('td').addClassName('name').update(res.name);
						tr.insert(td);
	
						// Description
						td = new Element('td').update(res.descr);
						tr.insert(td);
						
						// Insert delete button
						td = new Element('td');
						td.insert(delImg);

						// Add to cart button
						addImg = new Element('img', {src: '/media/themes/{{ user.get_profile.guistyle }}/img/misc/addtocart_small.gif'});
						addImg.c_data = $H(res);
						addImg.observe('click', function() {
							{{ plugin.id }}.addToCart(this.c_data);
						});
						td.insert(addImg);

						tr.insert(td);
						table.insert(tr);
					});
					div.insert(table);
				}
		);
	
		var post = {
			Plugin: uidskel,
			Method: 'getSavedItems'
		};
		r.send('/youpi/process/plugin/', $H(post).toQueryString());
	},

	/*
	 * Function: delSavedItem
	 * Delete a saved cart item
	 *
	 * Parameters:
	 *  trid - string: row id
	 *  name - string: item's name
	 *
	 */ 
	delSavedItem: function(trid, name) {
		var r = confirm("Are you sure you want to delete saved item '" + name + "'?");
		if (!r) return;

		var trNode = $(trid);
		var r = new HttpRequest(
				uidskel + '_result',
				null,	
				// Custom handler for results
				function(resp) {
					var node = trNode;
					var last = false;
					if (trNode.up('table').select('tr[id]').length == 1) {
						last = true;
						node = trNode.parentNode;
					}
		
					trNode.highlight({
						afterFinish: function() {
							node.fade({
								afterFinish: function() {
									node.remove();
									if (last) eval(uidskel + '.showSavedItems()');

									// Notify user
									document.fire('notifier:notify', "Item '" + name + "' successfully deleted");
								}
							});
						}
					});
				}
		);
	
		var post = {
			'Plugin': uidskel,
			'Method': 'deleteCartItem',
			'Name'	: name
		};

		r.send('/youpi/process/plugin/', $H(post).toQueryString());
	},

	/*
	 * Function: addToCart
	 * Add a skel item to cart
	 *
	 * Parameters:
	 *  data - hash: any user data
	 *
	 */ 
	addToCart: function(data) {
		var p_data = {	plugin_name : uidskel,
						userData : data
		};
	
		s_cart.addProcessing(p_data,
				// Custom hanlder
				function() {
					window.location.reload();
				}
		);
	}
};
